template argument substitution failed in arf.cc

```
arf.cc: In function 'std::ostream& exactreal::operator<<(std::ostream&, const exactreal::Arf&)':
arf.cc:190:15: error: no match for 'operator<<' (operand types are 'std::ostream' {aka 'std::basic_ostream<char>'} and 'int')
  190 |     return os << 0;
      |            ~~ ^~ ~
      |            |     |
      |            |     int
      |            std::ostream {aka std::basic_ostream<char>}
arf.cc:188:10: note: candidate: 'std::ostream& exactreal::operator<<(std::ostream&, const exactreal::Arf&)'
  188 | ostream& operator<<(ostream& os, const Arf& self) {
      |          ^~~~~~~~
arf.cc:188:45: note:   no known conversion for argument 2 from 'int' to 'const exactreal::Arf&'
  188 | ostream& operator<<(ostream& os, const Arf& self) {
      |                                  ~~~~~~~~~~~^~~~
In file included from /usr/include/c++/10.2.0/memory:84,
                 from ../exact-real/arf.hpp:24,
                 from arf.cc:21:
/usr/include/c++/10.2.0/bits/shared_ptr.h:69:5: note: candidate: 'template<class _Ch, class _Tr, class _Tp, __gnu_cxx::_Lock_policy _Lp> std::basic_ostream<_Ch, _Tr>& std::operator<<(std::basic_ostream<_Ch, _Tr>&, const std::__shared_ptr<_Tp, _Lp>&)'
   69 |     operator<<(std::basic_ostream<_Ch, _Tr>& __os,
      |     ^~~~~~~~
/usr/include/c++/10.2.0/bits/shared_ptr.h:69:5: note:   template argument deduction/substitution failed:
arf.cc:190:18: note:   mismatched types 'const std::__shared_ptr<_Tp, _Lp>' and 'int'
  190 |     return os << 0;
      |                  ^
In file included from /usr/include/c++/10.2.0/bits/basic_string.h:48,
                 from /usr/include/c++/10.2.0/string:55,
                 from /usr/include/gmpxx.h:41,
                 from ../exact-real/arf.hpp:28,
                 from arf.cc:21:
/usr/include/c++/10.2.0/string_view:622:5: note: candidate: 'template<class _CharT, class _Traits> std::basic_ostream<_Ch, _Tr>& std::operator<<(std::basic_ostream<_Ch, _Tr>&, std::basic_string_view<_CharT, _Traits>)'
  622 |     operator<<(basic_ostream<_CharT, _Traits>& __os,
      |     ^~~~~~~~
/usr/include/c++/10.2.0/string_view:622:5: note:   template argument deduction/substitution failed:
arf.cc:190:18: note:   mismatched types 'std::basic_string_view<_CharT, _Traits>' and 'int'
  190 |     return os << 0;
      |                  ^
In file included from /usr/include/c++/10.2.0/string:55,
                 from /usr/include/gmpxx.h:41,
                 from ../exact-real/arf.hpp:28,
                 from arf.cc:21:
/usr/include/c++/10.2.0/bits/basic_string.h:6458:5: note: candidate: 'template<class _CharT, class _Traits, class _Alloc> std::basic_ostream<_Ch, _Tr>& std::operator<<(std::basic_ostream<_Ch, _Tr>&, const std::__cxx11::basic_string<_CharT, _Traits, _Allocator>&)'
 6458 |     operator<<(basic_ostream<_CharT, _Traits>& __os,
      |     ^~~~~~~~
/usr/include/c++/10.2.0/bits/basic_string.h:6458:5: note:   template argument deduction/substitution failed:
arf.cc:190:18: note:   mismatched types 'const std::__cxx11::basic_string<_CharT, _Traits, _Allocator>' and 'int'
  190 |     return os << 0;
      |                  ^
In file included from /usr/include/boost/operators.hpp:97,
                 from ../exact-real/arf.hpp:29,
                 from arf.cc:21:
/usr/include/c++/10.2.0/cstddef:125:5: note: candidate: 'template<class _IntegerType> constexpr std::__byte_op_t<_IntegerType> std::operator<<(std::byte, _IntegerType)'
  125 |     operator<<(byte __b, _IntegerType __shift) noexcept
      |     ^~~~~~~~
/usr/include/c++/10.2.0/cstddef:125:5: note:   template argument deduction/substitution failed:
arf.cc:190:12: note:   cannot convert 'os' (type 'std::ostream' {aka 'std::basic_ostream<char>'}) to type 'std::byte'
  190 |     return os << 0;
      |            ^~
In file included from /usr/include/c++/10.2.0/bits/ios_base.h:46,
                 from /usr/include/c++/10.2.0/streambuf:41,
                 from /usr/include/c++/10.2.0/bits/streambuf_iterator.h:35,
                 from /usr/include/c++/10.2.0/iterator:66,
                 from /usr/include/boost/operators.hpp:98,
                 from ../exact-real/arf.hpp:29,
                 from arf.cc:21:
/usr/include/c++/10.2.0/system_error:262:5: note: candidate: 'template<class _CharT, class _Traits> std::basic_ostream<_Ch, _Tr>& std::operator<<(std::basic_ostream<_Ch, _Tr>&, const std::error_code&)'
  262 |     operator<<(basic_ostream<_CharT, _Traits>& __os, const error_code& __e)
      |     ^~~~~~~~
/usr/include/c++/10.2.0/system_error:262:5: note:   template argument deduction/substitution failed:
arf.cc:190:18: note:   cannot convert '0' (type 'int') to type 'const std::error_code&'
  190 |     return os << 0;
      |                  ^
```

--- a/libexactreal/src/arb.cc
+++ b/libexactreal/src/arb.cc
@@ -158,7 +158,7 @@ std::optional<bool> Arb::operator<(const Arb& rhs) const noexcept {
   } else if (arb_ge(arb_t(), rhs.arb_t())) {
     return false;
   } else {
-    return {};
+    return std::nullopt;
   }
 }
 
@@ -168,7 +168,7 @@ std::optional<bool> Arb::operator>(const Arb& rhs) const noexcept {
   } else if (arb_le(arb_t(), rhs.arb_t())) {
     return false;
   } else {
-    return {};
+    return std::nullopt;
   }
 }
 
@@ -177,7 +177,7 @@ std::optional<bool> Arb::operator<=(const Arb& rhs) const noexcept {
   if (gt.has_value()) {
     return !*gt;
   } else {
-    return {};
+    return std::nullopt;
   }
 }
 
@@ -186,7 +186,7 @@ std::optional<bool> Arb::operator>=(const Arb& rhs) const noexcept {
   if (lt.has_value()) {
     return !*lt;
   } else {
-    return {};
+    return std::nullopt;
   }
 }
 
@@ -196,7 +196,7 @@ std::optional<bool> Arb::operator==(const Arb& rhs) const noexcept {
   } else if (arb_ne(arb_t(), rhs.arb_t())) {
     return false;
   } else {
-    return {};
+    return std::nullopt;
   }
 }
 
@@ -205,7 +205,7 @@ std::optional<bool> Arb::operator!=(const Arb& rhs) const noexcept {
   if (eq) {
     return !*eq;
   } else {
-    return {};
+    return std::nullopt;
   }
 }
 

--- a/libexactreal/src/arf.cc
+++ b/libexactreal/src/arf.cc
@@ -22,6 +22,8 @@
 
 #include <arf.h>
 
+#include <ostream>
+
 using std::ostream;
 using std::string;
 

--- a/libexactreal/src/constraint_random_real_number.cc
+++ b/libexactreal/src/constraint_random_real_number.cc
@@ -80,7 +80,7 @@ class ConstraintRandomRealNumber final : public RealNumber {
   }
 
   explicit operator std::optional<mpq_class>() const override {
-    return {};
+    return std::nullopt;
   }
 
   RealNumber const& operator>>(ostream& out) const override {

--- a/libexactreal/src/element.cc
+++ b/libexactreal/src/element.cc
@@ -303,13 +303,13 @@ std::optional<Element<Ring>> Element<Ring>::truediv(const Element<Ring>& rhs) co
 
     auto generator_quotient = *g / *h;
     if (!generator_quotient.has_value())
-      return {};
+      return std::nullopt;
 
     auto coefficient_quotient = a / b;
 
     if (coefficient_quotient * b != a) {
       ASSERT(!Ring::isField, "division of coefficients must be exact in fields");
-      return {};
+      return std::nullopt;
     }
 
     auto partial_quotient = coefficient_quotient * Module<Ring>::make({*generator_quotient}, this->module()->ring())->gen(0);

--- a/libexactreal/src/element.cc
+++ b/libexactreal/src/element.cc
@@ -110,7 +110,7 @@ Element<Ring>::Element(const Element<IntegerRing>& value)
     : impl(spimpl::make_impl<Element<Ring>::Implementation>(
           Module<Ring>::make(value.module()->basis()),
           [&](const auto& coefficients) {
-            return std::vector<typename Ring::ElementClass>(coefficients.begin(), coefficients.end());
+            return std::vector<typename Ring::ElementClass>(begin(coefficients), end(coefficients));
           }(value.coefficients()))) {}
 
 template <typename Ring>
@@ -119,7 +119,7 @@ Element<Ring>::Element(const Element<RationalField>& value)
     : impl(spimpl::make_impl<Element<Ring>::Implementation>(
           Module<Ring>::make(value.module()->basis()),
           [&](const auto& coefficients) {
-            return std::vector<typename Ring::ElementClass>(coefficients.begin(), coefficients.end());
+            return std::vector<typename Ring::ElementClass>(begin(coefficients), end(coefficients));
           }(value.coefficients()))) {}
 
 template <typename Ring>
@@ -221,8 +221,8 @@ Element<Ring>& Element<Ring>::operator*=(const Element<Ring>& rhs) {
     }
   }
 
-  vector<std::pair<shared_ptr<const RealNumber>, typename Ring::ElementClass>> sorted(products.begin(), products.end());
-  std::sort(sorted.begin(), sorted.end(), [](const auto& lhs, const auto& rhs) { return lhs.first->deglex(*rhs.first); });
+  vector<std::pair<shared_ptr<const RealNumber>, typename Ring::ElementClass>> sorted(begin(products), end(products));
+  std::sort(begin(sorted), end(sorted), [](const auto& lhs, const auto& rhs) { return lhs.first->deglex(*rhs.first); });
 
   vector<shared_ptr<const RealNumber>> basis;
   vector<typename Ring::ElementClass> coefficients;
@@ -497,11 +497,11 @@ bool Element<Ring>::operator==(const Element<Ring>& rhs) const {
 template <typename Ring>
 bool Element<Ring>::operator==(const RealNumber& rhs) const {
   auto& gens = impl->parent->basis();
-  auto it = find_if(gens.begin(), gens.end(), [&](const auto& other) { return *other == rhs; });
-  if (it == gens.end()) {
+  auto it = find_if(begin(gens), end(gens), [&](const auto& other) { return *other == rhs; });
+  if (it == end(gens)) {
     throw logic_error("not implemented - equality of Element with unrelated RealNumber");
   }
-  const size at = it - gens.begin();
+  const size at = it - begin(gens);
   for (size_t i = 0; i < impl->parent->rank(); i++) {
     if (impl->coefficients[i] == 0) {
       if (i == at) {
@@ -520,7 +520,7 @@ bool Element<Ring>::operator==(const RealNumber& rhs) const {
 
 template <typename Ring>
 Element<Ring>::operator bool() const {
-  return std::any_of(impl->coefficients.begin(), impl->coefficients.end(), [](auto c) { return c != 0; });
+  return std::any_of(begin(impl->coefficients), end(impl->coefficients), [](auto c) { return c != 0; });
 }
 
 template <typename Ring>

--- a/libexactreal/src/module.cc
+++ b/libexactreal/src/module.cc
@@ -52,8 +52,8 @@ class Module<Ring>::Implementation {
   Implementation() : basis({}), parameters(trivial()) {}
 
   explicit Implementation(const Basis& basis, const Ring& parameters) : basis(basis), parameters(parameters) {
-    for (auto it = basis.begin(); it != basis.end(); it++) {
-      for (auto jt = it + 1; jt != basis.end(); jt++) {
+    for (auto it = begin(basis); it != end(basis); it++) {
+      for (auto jt = it + 1; jt != end(basis); jt++) {
         if (static_cast<std::optional<mpq_class>>(**it) && static_cast<std::optional<mpq_class>>(**jt)) {
           throw std::logic_error("at most one generator can be rational");
         }
@@ -142,7 +142,7 @@ shared_ptr<const Module<Ring>> Module<Ring>::span(const shared_ptr<const Module<
 
   auto basis = m->basis();
   for (auto& ngen : n->basis()) {
-    if (find_if(basis.begin(), basis.end(), [&](const auto& gen) { return *gen == *ngen; }) == basis.end()) {
+    if (find_if(begin(basis), end(basis), [&](const auto& gen) { return *gen == *ngen; }) == end(basis)) {
       basis.push_back(ngen);
     }
   }
@@ -154,7 +154,7 @@ template <typename Ring>
 bool Module<Ring>::submodule(const Module<Ring>& supermodule) const {
   auto super = supermodule.basis();
   for (auto& gen : basis()) {
-    if (find_if(super.begin(), super.end(), [&](const auto& sgen) { return *gen == *sgen; }) == super.end()) {
+    if (find_if(begin(super), end(super), [&](const auto& sgen) { return *gen == *sgen; }) == end(super)) {
       return false;
     }
   }

--- a/libexactreal/src/number_field.cc
+++ b/libexactreal/src/number_field.cc
@@ -46,7 +46,7 @@ std::optional<mpq_class> NumberField::rational(const ElementClass& x) {
   return x.is_rational() ? std::optional{static_cast<mpq_class>(x)} : std::nullopt;
 }
 
-bool NumberField::operator==(const NumberField& rhs) const { return parameters == rhs.parameters; }
+bool NumberField::operator==(const NumberField& rhs) const { return *parameters == *rhs.parameters; }
 
 mpz_class NumberField::floor(const ElementClass& x) { return x.floor(); }
 

--- a/libexactreal/src/random_real_number.cc
+++ b/libexactreal/src/random_real_number.cc
@@ -81,7 +81,7 @@ class RandomRealNumber final : public RealNumber {
   }
 
   operator std::optional<mpq_class>() const override {
-    return {};
+    return std::nullopt;
   }
 
   RealNumber const& operator>>(ostream& out) const override {

--- a/libexactreal/src/real_number.cc
+++ b/libexactreal/src/real_number.cc
@@ -87,7 +87,7 @@ std::optional<std::shared_ptr<const RealNumber>> RealNumber::operator/(const Rea
     throw std::logic_error("not implemented: division of generic real number by rational");
   }
 
-  return {};
+  return std::nullopt;
 }
 
 bool RealNumber::operator<(const RealNumber& rhs) const {

--- a/libexactreal/src/real_number_product.cc
+++ b/libexactreal/src/real_number_product.cc
@@ -72,7 +72,7 @@ class RealNumberProduct final : public RealNumber {
   }
 
   explicit operator std::optional<mpq_class>() const override {
-    return {};
+    return std::nullopt;
   }
 
   std::optional<std::shared_ptr<const RealNumber>> operator/(const RealNumber& rhs) const override {
@@ -99,7 +99,7 @@ class RealNumberProduct final : public RealNumber {
         if (quotient[d.first] == 0)
           quotient.erase(d.first);
         else if (quotient[d.first] < 0)
-          return {};
+          return std::nullopt;
       }
     }
 

--- a/libexactreal/src/real_number_product.cc
+++ b/libexactreal/src/real_number_product.cc
@@ -53,8 +53,8 @@ auto& factory() {
 class RealNumberProduct final : public RealNumber {
  public:
   explicit RealNumberProduct(const Factors& factors) : factors(factors) {
-    ASSERT(std::all_of(factors.begin(), factors.end(), [](auto& factor) { return factor.second >= 1; }), "factors must appear at least once");
-    ASSERT(std::all_of(factors.begin(), factors.end(), [](auto& factor) { return !static_cast<std::optional<mpq_class>>(*factor.first); }), "factors must be transcendental");
+    ASSERT(std::all_of(begin(factors), end(factors), [](auto& factor) { return factor.second >= 1; }), "factors must appear at least once");
+    ASSERT(std::all_of(begin(factors), end(factors), [](auto& factor) { return !static_cast<std::optional<mpq_class>>(*factor.first); }), "factors must be transcendental");
   }
 
   RealNumber const& operator>>(std::ostream& os) const override {
@@ -87,7 +87,7 @@ class RealNumberProduct final : public RealNumber {
 
     Factors quotient = factors;
 
-    if (quotient.find(rhs.shared_from_this()) != quotient.end()) {
+    if (quotient.find(rhs.shared_from_this()) != end(quotient)) {
       auto rhs_ = rhs.shared_from_this();
       quotient[rhs_]--;
       if (quotient[rhs_] == 0)
@@ -247,9 +247,9 @@ bool RealNumber::deglex(const RealNumber& rhs_) const {
 
       for (auto& gen : gens) {
         ASSERT(!static_cast<std::optional<mpq_class>>(*gen).has_value(), "factors of real number product must not be rational");
-        if (lhs->factors.find(gen) == lhs->factors.end())
+        if (lhs->factors.find(gen) == end(lhs->factors))
           return false;
-        if (rhs->factors.find(gen) == rhs->factors.end())
+        if (rhs->factors.find(gen) == end(rhs->factors))
           return true;
         if (lhs->factors.at(gen) < rhs->factors.at(gen))
           return false;
@@ -270,7 +270,7 @@ void load_product(cereal::JSONInputArchive& archive, std::shared_ptr<const RealN
   Factors factors;
   archive(cereal::make_nvp("factors", factors));
   base = factory().get(factors, [&]() {
-    return new RealNumberProduct(Factors(factors.begin(), factors.end()));
+    return new RealNumberProduct(Factors(begin(factors), end(factors)));
   });
 }
 }  // namespace exactreal

--- a/libexactreal/test/cereal.test.cc
+++ b/libexactreal/test/cereal.test.cc
@@ -47,7 +47,7 @@ template <typename T>
 std::string toString(const T& x) {
   if constexpr (is_shared_ptr<T>::value) {
     if (x)
-      return boost::lexical_cast<std::string>(x) + "->" + boost::lexical_cast<std::string>(*x);
+      return boost::lexical_cast<std::string>(&*x) + "=&" + boost::lexical_cast<std::string>(*x);
     else
       return "null";
   } else if constexpr (is_vector<T>::value) {

--- a/libexactreal/test/module.test.cc
+++ b/libexactreal/test/module.test.cc
@@ -35,6 +35,11 @@ using std::vector;
 namespace exactreal::test {
 
 TEST_CASE("Module over ZZ", "[module][integer_ring]") {
+  SECTION("Uniqueness") {
+    REQUIRE(&*Module<IntegerRing>::make({}) == &*Module<IntegerRing>::make({}));
+    REQUIRE(&*Module<IntegerRing>::make({RealNumber::random()}) != &*Module<IntegerRing>::make({RealNumber::random()}));
+  }
+
   SECTION("Rank") {
     auto trivial = Module<IntegerRing>::make({});
     REQUIRE(trivial->rank() == 0);
@@ -42,7 +47,7 @@ TEST_CASE("Module over ZZ", "[module][integer_ring]") {
 
     auto m = Module<IntegerRing>::make({RealNumber::random(), RealNumber::random()});
     REQUIRE(m->rank() == 2);
-    REQUIRE(lexical_cast<string>(*m) == "ℤ-Module(ℝ(0.120809…), ℝ(0.303644…))");
+    REQUIRE(lexical_cast<string>(*m) == "ℤ-Module(ℝ(0.178808…), ℝ(0.478968…))");
 
     auto n = Module<IntegerRing>::make(m->basis());
     REQUIRE(m == n);

--- a/news/cereal.rst
+++ b/news/cereal.rst
@@ -1,0 +1,3 @@
+**Fixed:**
+
+* deserialization of modules over number fields; shared pointers to renf_class was not compared correctly which lead to a module not being identical to a deserialization of a serialization of itself; part of #106.

--- a/news/nullopt.rst
+++ b/news/nullopt.rst
@@ -1,0 +1,3 @@
+**Fixed:**
+
+* some minor C++ coding style such as nullopt instead of {}

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -50,7 +50,7 @@ requirements:
   host:
 {%- if libexactreal_requirements %}
     - boost-cpp
-    - e-antic >=1.0a master_1148
+    - e-antic >=1.0a1
     - arb
     - gmp
     - gmpxxll
@@ -67,7 +67,7 @@ requirements:
     # enable test/pyexactreal tests in ./configure
     - pytest
     - cppyy
-    - pyeantic >=1.0.1
+    - pyeantic >=1.1.1
     # enable test/libexactreal tests in ./configure
     - benchmark
     # enable check-valgrind in ./configure
@@ -84,7 +84,7 @@ requirements:
 {%- if component == "pyexactreal" %}
     - python {{ _python }}
     - boost-cpp
-    - pyeantic >=1.0.1
+    - pyeantic >=1.1.1
     - cppyy
 {%- endif %}
 

